feat(webvh): Added did:webvh resolver

brianorwhatever opened the pull request:
@genaris @TimoGlastra as discussed I have pulled the crypto out of didwebvh-ts while also removing all of the dependencies. Please take a look at this new PR

TimoGlastra commented on packages/webvh/src/WebvhModuleConfig.ts:
```diff
@@ -0,0 +1,22 @@
+
+  /** See {@link WebvhModuleConfigOptions.baseUrl} */
+  public get baseUrl() {
+    return this.options.baseUrl ?? 'https://webvh.io'
```
What does this default domain do? Is the baseUrl for when you create DIDs? I think having no default value and allowing it to be passed when creating one makes sense.

TimoGlastra commented on packages/webvh/src/dids/WebvhDidRegistrar.ts:
```diff
@@ -0,0 +1,239 @@
+          }
+        }
+        const privateKey = Buffer.from(MultiBaseEncoder.decode(privateKeyMultibase).data)
+        if (privateKeyMultibase && !isValidPrivateKey(privateKey, KeyType.Ed25519)) {
```
Is Ed25519 required? Just curious, we'd like to use it with P256 as well

brianorwhatever replied:
Yes unfortunately ED25519 is required for did:webvh for update keys. We did it that way to standardize it on one interoperable crypto suite you can of course use any keys in a DID document you desire.

TimoGlastra commented on packages/webvh/src/dids/WebvhDidRegistrar.ts:
```diff
@@ -0,0 +1,239 @@
+            },
+          }
+        }
+        const privateKey = Buffer.from(MultiBaseEncoder.decode(privateKeyMultibase).data)
```
Is it possible to pass an existing public key, but not the private key? We usually create a key first, then use that to create a diddoc

(This is fine to add later, just trying to understand how it works)

TimoGlastra commented on packages/webvh/src/dids/WebvhDidRegistrar.ts:
```diff
@@ -0,0 +1,239 @@
+      }
+
+      // Register the DID using didwebvh-ts
+      const response = await createDID()
```
Shouldn't we somehow pass the did document here?

TimoGlastra commented on packages/webvh/src/dids/WebvhDidRegistrar.ts:
```diff
@@ -0,0 +1,239 @@
+      }
+
+      // Update the DID using didwebvh-ts
+      const response = await updateDID({doc: options.didDocument.toJSON()})
```
I think we should first fetch the record to see if it exists

TimoGlastra commented on packages/webvh/src/dids/WebvhDidRegistrar.ts:
```diff
@@ -0,0 +1,239 @@
+  }
+
+  public async deactivate(agentContext: AgentContext, options: WebvhDidDeactivateOptions): Promise<DidDeactivateResult> {
+    // const didRepository = agentContext.dependencyManager.resolve(DidRepository)
```
If this is not supported yet we can return an error for now and remove this code

TimoGlastra commented on packages/webvh/src/dids/WebvhDidRegistrar.ts:
```diff
@@ -0,0 +1,239 @@
+}
+
+interface IVerificationMethod {
+  type: string
```
Type must be MultiKey right?

TimoGlastra commented on packages/webvh/src/dids/didWebvhUtil.ts:
```diff
@@ -0,0 +1,94 @@
+    throw new Error('Not implemented')
+  }
+
+  public async verify(signature: Uint8Array, message: Uint8Array, publicKey: Uint8Array): Promise<boolean> {
```
It's probably good if the interface from did web vh passes the keytype as well, so it's ready to support additional key types as well

TimoGlastra commented on packages/webvh/src/dids/didWebvhUtil.ts:
```diff
@@ -0,0 +1,94 @@
+        signature: Buffer.from(signature),
+      })
+    } catch (error) {
+      // Log error in a non-production environment
```
You can use agentContext.config.logger.error for this, and can be disabled by the user of the framework

TimoGlastra commented on packages/webvh/src/dids/didWebvhUtil.ts:
```diff
@@ -0,0 +1,94 @@
+export class DIDWebvhCrypto extends AbstractCrypto {
+  private agentContext?: AgentContext
+
+  public constructor(agentContext?: AgentContext) {
```
I think we can make agent context required

TimoGlastra reviewed (commented):
The new approach is awesome, thanks @brianorwhatever. I left some comments. It seems some parts are todo, do you want to address these as part of this PR, or in another one?

brianorwhatever commented:
@TimoGlastra I forgot to mention I was focussing solely on resolver in this PR. I added some of the base stuff for the registrar, which will be coming soon. I'm happy to either remove that stuff or fix it up later. Higher priority is resolving DID documents and resources.

TimoGlastra commented:
Ah that makes sense @brianorwhatever. Let's focus this pr on the resolving in that case and we can add the registration parts next 👍

brianorwhatever commented:
OK, great I will remove the stuff about the registrar. Unfortunately, I don't think that this is 0.3 compatible I can compare the differences and see what would be required to make it work.

brianorwhatever commented:
Hey @TimoGlastra, quick question about verifying Data Integrity proofs in a custom setup.

We're building an `AnonCredsRegistry` for `did:webvh` (`WebVhAnonCredsRegistry`). The objects we fetch (schemas, cred defs, etc.) are structured as simple JSON-LD "AttestedResources" which basically wrap the `content` (the actual AnonCreds object) and include a standard DI `proof` (e.g., `eddsa-jcs-2022`).

Inside our registry logic (in a helper `_resolveAndValidateAttestedResource`), we're already handling:

*   Resolving the `did:webvh:.../resource/hash` URI.
*   Parsing the JSON-LD.
*   Checking the `type` includes `AttestedResource`.
*   Verifying a multihash of the canonicalized `content` matches the `hash` in the URI.
*   Checking the `issuerId` in the `content` matches the DID from the URI.

The last piece is actually verifying the `proof` signature itself. We looked into `DataIntegrityService` and `W3cCredentialService`, but they didn't seem designed for verifying proofs on arbitrary signed objects like this (it's not strictly a VC/VP).

We also considered using the signature suites directly (like `Ed25519Signature2020`), getting them from the `SignatureSuiteRegistry`, and calling their `verifyProof` method. This looks possible but feels like we might be duplicating internal logic, especially around preparing the data correctly for the specific suite and ensuring the `documentLoader` (which uses our `WebvhDidResolver`) is passed through correctly to resolve the `verificationMethod`.

**Is there a recommended/simpler way in Credo to verify a DI proof on a custom JSON-LD object like this, making sure key resolution via the `verificationMethod` works as expected?** Or is using the signature suite directly the intended path here?

Currently have it stubbed out, but want to make sure we implement the crypto verification correctly using Credo's infrastructure.

Thanks!

brianorwhatever commented:
@TimoGlastra @genaris this is ready for a review. I have updated the didwebvh-ts package to the new version which supports version 1.0 of the spec and addressed any comments in this thread. Thanks!

TimoGlastra commented on packages/webvh/package.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "@credo-ts/webvh",
+  "version": "0.4.0",
```
Should be synced with the other packages

```suggestion
  "version": "0.5.13",
```

TimoGlastra commented on packages/webvh/src/anoncreds/utils/multihash.ts:
We have a multihash implementation already: https://github.com/openwallet-foundation/credo-ts/blob/879ed2c9178684d14e6850cbcdb02d6832e0d730/packages/core/src/utils/MultiHashEncoder.ts.

If there's any features missing in our multi hash implementation, it would be better to extend the existing functionaility, this way we prevent duplication of logic

TimoGlastra commented on packages/webvh/src/anoncreds/utils/transform.ts:
```diff
@@ -0,0 +1,161 @@
+  @Type(() => String)
+  public attrNames!: string[]
+
+  @IsString() // Example validation
```
"Example validation" can be removed everywhere i think?

TimoGlastra commented on packages/webvh/src/dids/WebvhDidResolver.ts:
```diff
@@ -0,0 +1,128 @@
+
+export class WebvhDidResolver implements DidResolver {
+  public readonly supportedMethods = ['webvh']
+  public readonly allowsCaching = false
+  public readonly allowsLocalDidRecord = false
```
Any reason these are both false? For the `did:web` one they are both true, I'd assume the same values here

TimoGlastra commented on packages/webvh/src/dids/WebvhDidResolver.ts:
```diff
@@ -0,0 +1,128 @@
+    try {
+      return await this.resolveDidDoc(agentContext, did)
+    } catch (error) {
+      agentContext.config.logger.error(`Error resolving DID: ${error}`)
```
```suggestion
      agentContext.config.logger.error(`Error resolving DID ${did}: ${error}`)
```

TimoGlastra commented on packages/webvh/src/dids/WebvhDidResolver.ts:
```diff
@@ -0,0 +1,128 @@
+      agentContext.config.logger.debug(`Fetching resource from: ${httpsUrl}`)
+      
+      // Fetch the resource using native fetch
+      const response = await fetch(httpsUrl)
```
Should use fetch from the agent config. `agentContext.config.agentDependencies.fetch`

TimoGlastra commented on packages/webvh/src/dids/didWebvhUtil.ts:
```diff
@@ -0,0 +1,119 @@
+
+  public constructor(agentContext: AgentContext) {
+    super({
+      verificationMethod: {
+        id: 'did:webvh:123',
+        controller: 'did:webvh:123',
+        type: 'Ed25519VerificationKey2020',
+        publicKeyMultibase: '123',
+        secretKeyMultibase: '123',
+      },
```
Why do we set it to these sample values? Should this be optional in the `AbstractCrypto`?

TimoGlastra commented on packages/webvh/src/dids/didWebvhUtil.ts:
```diff
@@ -0,0 +1,119 @@
+  }
+}
+
+export interface IDidDocOptions {
```
Where is this interface used? We generally don't use `I` prefix for interfaces, and also write out the full name, so `DidDocumentOptions`, but might be good to make it more specific if it's related to WebVh.

TimoGlastra commented on packages/webvh/tests/webvh-did-resolver.test.ts:
```diff
@@ -0,0 +1,75 @@
+    agent = new Agent({...agentOptions, modules: { ...getWebvhModules(), inMemory: new InMemoryWalletModule() }})
+
+    const initPromise = agent.initialize()
+    const timeoutPromise = new Promise((_, reject) =>
+      setTimeout(() => reject(new Error('Agent initialization timed out')), 5000)
+    )
+
+    await Promise.race([initPromise, timeoutPromise])
```
Any reason for this approach? Since you mock the resolve call, i don't see any reason why initialize would time out?

TimoGlastra commented on packages/webvh/tests/webvh-did-resolver.test.ts:
```diff
@@ -0,0 +1,75 @@
+  })
+
+  it('should resolve a valid WebVH DID', async () => {
+    let err, res
+    try {
+      const didResolutionResult = await agent.dids.resolve(validDid)
+      res = didResolutionResult
+    } catch (e) {
+      err = e
+    }
+
+    expect(err).toBeUndefined()
+    expect(res?.didDocument).toBeDefined()
+    expect(res?.didDocument?.id).toBe('did:webvh:QmdmPkUdYzbr9txmx8gM2rsHPgr5L6m3gHjJGAf4vUFoGE:domain.example')
```
```suggestion
      const didResolutionResult = await agent.dids.resolve(validDid)
    expect(didResolutionResult.didDocument).toBeDefined()
 expect(didResolutionResult.didDocument?.id).toBe('did:webvh:QmdmPkUdYzbr9txmx8gM2rsHPgr5L6m3gHjJGAf4vUFoGE:domain.example')
```

I think this tests exactly the same?

TimoGlastra commented on packages/webvh/tsconfig.build.json:
```diff
@@ -0,0 +1,13 @@
+    "outDir": "./build",
+    "baseUrl": ".",
+    "skipDefaultLibCheck": true,
+    "paths": {
+      "@credo-ts/core": ["../core/build"],
+      "@credo-ts/core/*": ["../core/build/*"]
+    }
```
this should not be needed

TimoGlastra commented on packages/webvh/src/anoncreds/services/WebVhAnonCredsRegistry.ts:
```diff
@@ -0,0 +1,568 @@
+      return true
+
+    } catch (error) {
+      agentContext.config.logger.error('Error during proof validation', {
```
For all error messages here it would be useful to be more specific. If you see this in the context of the full credo agent, it can be hard to understand where a message originates from

```suggestion
      agentContext.config.logger.error('Error during did:webvh resource proof validation', {
```

brianorwhatever commented on packages/webvh/src/anoncreds/services/__tests__/WebVhAnonCredsRegistry.test.ts:
```diff
@@ -0,0 +1,681 @@
+    mockVerify.mockReset()
+
+    const agentConfig = getAgentConfig('WebVhAnonCredsRegistryTest')
+    agentContext = getAgentContext({ agentConfig })
+    
+    // Register the mocked resolver instance
+    agentContext.dependencyManager.registerInstance(WebvhDidResolver, new WebvhDidResolver())
+    
+    // Mock the dependency manager resolve method
+    const originalResolve = agentContext.dependencyManager.resolve.bind(agentContext.dependencyManager)
+    agentContext.dependencyManager.resolve = jest.fn().mockImplementation((token) => {
+      const tokenString = token?.name || token?.toString?.() || String(token)
+      
+      if (tokenString.includes('DidsApi')) {
+        return mockDidsApi
+      }
+      if (tokenString.includes('KeyManagementApi')) {
+        return mockKeyManagementApi
+      }
+      if (token === WebvhDidResolver || tokenString.includes('WebvhDidResolver')) {
+        return { resolveResource: mockResolveResource }
+      }
+      return originalResolve(token)
+    })
```
worked great. thanks!

TimoGlastra commented on packages/webvh/src/dids/didWebvhUtil.ts:
```diff
@@ -0,0 +1,119 @@
+    }
+  }
+
+  private async verifyWithLegacyMethod(signature: Uint8Array, message: Uint8Array, publicKey: Uint8Array): Promise<boolean> {
```
Yes the main branch is for 0.6.0. 

If we want this to be available in both i think the best method is to:
- remove all legacy KMs stuff from this PR
- merge into main
- create a PR to 0.5.x branch, with changes for the legacy KMS

But when targeting the main branch, we should only use the new KMS API

brianorwhatever replied:
Perfect, that makes sense. I'm in the process of removing the legacy stuff and I think I have an idea of how to get 0.5 working. Just created a PR for didwebvh-ts that removes the requirement of passing a VM into the constructor so once that lands I'll have this PR in great shape

TimoGlastra commented on packages/webvh/package.json:
```diff
@@ -0,0 +1,39 @@
+  "dependencies": {
+    "@credo-ts/anoncreds": "workspace:*",
+    "@credo-ts/core": "workspace:*",
+    "@multiformats/base-x": "^4.0.1",
```
I think this can be removed now?

TimoGlastra commented on packages/webvh/src/anoncreds/services/WebVhAnonCredsRegistry.ts:
```diff
@@ -0,0 +1,552 @@
+      })
+
+      let errorType: GetSchemaReturn['resolutionMetadata']['error'] = 'internalError'
+      if (error instanceof CredoError) {
+        if (error.message.includes('parse resource data')) errorType = 'invalidJson'
+        else if (error.message.includes('resolve') || error.message.includes('missing data')) errorType = 'notFound'
+        else if (error.message.includes('attested')) errorType = 'invalid'
+        else if (error.message.includes('proof')) errorType = 'invalid'
+        else if (error.message.includes('hash mismatch')) errorType = 'invalid'
+        else if (error.message.includes('not a valid schema')) errorType = 'invalid'
+        else if (error.message.includes('Invalid schema ID')) errorType = 'invalidDid'
+        else if (error.message.includes('Issuer ID mismatch')) errorType = 'invalid'
+        else if (error.message.includes('missing a valid issuerId')) errorType = 'invalid'
+      }
```
Where are these error messsages constructed? We don't use error messages anywhere to handle errors as it's very brittle. If you want to catch these cases, best i think is to create and throw specific errors that you can check with instanceof

brianorwhatever commented:
@TimoGlastra I think this is ready to be looked at again now. Working on a 0.5.x version now

brianorwhatever commented:
legacy version created in #2339

swcurran commented:
@genaris @TimoGlastra @cvarjao @berendsliedrecht -- can someone please review this and the #2339 PR?  We would really like to get this into Bifold and beyond.

genaris reviewed (approved):
Great job here! LGTM

Looking forward to work also in the DID and resource creation/update!

genaris commented:
@brianorwhatever are you able to address the two minor changes I've spotted in my review? I wanted to do it myself but I cannot commit to your branch because it is in an organization fork.

brianorwhatever commented on packages/webvh/src/dids/WebvhDidResolver.ts:
```diff
@@ -0,0 +1,128 @@
+      agentContext.config.logger.debug(`Fetching resource from: ${httpsUrl}`)
+      
+      // Fetch the resource using native fetch
+      const response = await fetch(httpsUrl)
```
resolved this in both PRs

brianorwhatever commented:
@genaris I have updated both this and #2339 with the fetch dependency from agent context

genaris commented on packages/webvh/package.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "@credo-ts/webvh",
+  "version": "0.6.0",
```
I understand, but all other modules in main branch do have this version. Please update it to 0.5.13 so all of them are in sync.

brianorwhatever replied:
ah ok I understand now. Done!

genaris commented:
@brianorwhatever I've created a PR in your branch to fix some remaining issues regarding the types: https://github.com/aviarytech/credo-ts/pull/2

brianorwhatever commented:
thanks @genaris!
